src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
 ﻿using System.Collections.Generic;
-
+using System.Linq;
 using Dynamo.Models;
+using Dynamo.Nodes;
+using System;
+using ProtoCore.AST.AssociativeAST;
+using ProtoCore.DSASM;
+using Dynamo.Utilities;
 
 namespace Dynamo.DSEngine
 {
-    public static class NodeToCodeUtils
+    /// <summary>
+    /// The result of converting nodes to code. As when a node is converted to 
+    /// code, its inputs and outputs may be renamed to avoid confliction, the
+    /// result contains maps for inputs/new-inputs and outputs/new-outputs.
+    /// </summary>
+    public class NodeToCodeResult
+    {
+        /// <summary>
+        /// AST nodes that compiled from NodeModel.
+        /// </summary>
+        public IEnumerable<AssociativeNode> AstNodes { get; private set; }
+
+        /// <summary>
+        /// The map between original input name and new name.
+        /// </summary>
+        public Dictionary<string, string> InputMap { get; private set; }
+
+        /// <summary>
+        /// The map between original output name and new name.
+        /// </summary>
+        public Dictionary<string, string> OutputMap { get; private set; }
+
+        public NodeToCodeResult(
+            IEnumerable<AssociativeNode> astNodes, 
+            Dictionary<string, string> inputMap, 
+            Dictionary<string, string> outputMap)
+        {
+            AstNodes = astNodes;
+            InputMap = inputMap;
+            OutputMap = outputMap;
+        }
+    }
+    public class NodeToCodeUtils
     {
-        public static string ConvertNodesToCode(AstBuilder astBuilder, IEnumerable<NodeModel> nodeList, bool verboseLogging)
+        /// <summary>
+        /// Traverse all identifiers in depth-first order and for each 
+        /// identifier apply a function to it.
+        /// </summary>
+        private class IdentifierVisitor
+        {
+            public static void Visit(AssociativeNode astNode, Action<IdentifierNode> func)
+            {
+                if (astNode is IdentifierNode)
+                {
+                    var identNode = astNode as IdentifierNode;
+                    func(identNode);
+                    Visit(identNode.ArrayDimensions, func);
+                }
+                else if (astNode is IdentifierListNode)
+                {
+                    var node = astNode as IdentifierListNode;
+                    Visit(node.LeftNode, func);
+                    Visit(node.RightNode, func);
+                }
+                else if (astNode is FunctionCallNode)
+                {
+                    var node = astNode as FunctionCallNode;
+                    for (int i = 0; i < node.FormalArguments.Count; ++i)
```

Comment:
Okay got it, the visitor function may change the content of `Name` or `Value`, and possibly `FormalArguments` too.